File: rdvhome/switches/nanoleaf.py
# -*- coding: utf-8 -*-

from __future__ import absolute_import, print_function, unicode_literals

import aiohttp

from rpy.functions.datastructures import data

from rdvhome.switches.base import capabilities
from rdvhome.switches.philips import RemoteBase, debounce, remove_none
from rdvhome.utils import json
from rdvhome.utils.colors import (
    HSB, color_to_homekit, color_to_nanoleaf, color_to_philips, color_to_homekit,
    homekit_to_color, philips_to_color, to_color
)


class NanoleafControl(RemoteBase):

    def __init__(self, id, effects=None, **opts):
        self.effects = data(effects or {})
        super().__init__(id, **opts)

    @property
    def default_capabilities(self):
        return capabilities(
            on=True,
            hue=True,
            saturation=True,
            brightness=True,
            effects=self.effects,
        )

    def get_api_url(self, path="/"):
        return "http://%s:16021/api/v1/%s%s" % (self.ipaddress, self.access_token, path)

    async def api_request(self, path="", payload=None):
        # Nanoleaf answers writes with an empty 204 body (and an empty 400 for
        # an unknown effect), so parse the response only when there is one.
        url = self.get_api_url(path)

        async with aiohttp.ClientSession() as session:
            method = session.put(url, json=payload) if payload else session.get(url)
            async with method as response:
                text = await response.text()
                return json.loads(text) if text.strip() else data()

    @debounce(1)
    async def get_nanoleaf_status(self):
        state = await self.api_request('/state')
        effect = await self.api_request('/effects/select')

        return data(
            on=state.on.value,
            allow_on=True,
            hue=state.hue.value / state.hue.max,
            brightness=state.brightness.value / state.brightness.max,
            saturation=state.sat.value / state.sat.max,
            effect=effect if effect in self.effects else None,
        )

    async def status(self):
        defaults = await self.get_nanoleaf_status()
        return await self.send(**defaults)

    def _get_state_changes(self, on, color):

        if on is not None:
            yield "on", {"value": on}

        if color is not None and on is not False:
            for key, value in color_to_nanoleaf(color).items():
                yield key, {"value": value}

    async def switch(self, on=None, color=None, effect = None, **opts):


        if effect:

            if isinstance(effect, str):
                # Scenes broadcast the same effect to every nanoleaf, but each
                # device exposes a different set, so ignore one we don't have.
                if self.effects and effect not in self.effects:
                    return await self.send()

                await self.api_request('/effects', payload = {'select': effect})
                return await self.send(on = True, color = color, effect = effect)

            else:
                await self.api_request('/effects', payload = {'write': {
                    "command": "display",
                    "animName": "New animation",
                    "animType": "highlight",
                    "colorType": "HSB",
                    "animData": None,
                    "palette": tuple(color_to_homekit(dict(brightness = 1, hue = effect.hue, saturation = effect.saturation / (i+1))) for i in range(3)),
                    "brightnessRange": {
                        "minValue": 50,
                        "maxValue": 100
                    },
                    "transTime": {
                        "minValue": 5,
                        "maxValue": 10
                    },
                    "delayTime": {
                        "minValue": 5,
                        "maxValue": 10
                    },
                    "loop": True
                }})
                return await self.send(effect = None, on = True, color = effect)

        defaults = dict(self._get_state_changes(on, color))

        await self.api_request('/state', payload = defaults)

        # Setting colour/power leaves the device in solid mode: clear any
        # previously selected effect so the UI stops highlighting it.
        return await self.send(effect = None, **remove_none(on=on, color = color))

    async def is_on(self):
        return (await self.get_nanoleaf_status()).on

File: rdvhome-rs/src/device/nanoleaf.rs
// Nanoleaf panels. No polling: the state is asked to the panel on every read.

use std::sync::Arc;

use async_trait::async_trait;
use serde_json::{json, Map, Number, Value};

use super::{Capabilities, ColorReport, Command, Device, Effect, Report};
use crate::color::{Hsb, HOMEKIT};

pub struct Nanoleaf {
    client: reqwest::Client,
    // http://<host>:16021/api/v1/<token>
    base: String,
    effects: Vec<String>,
}

impl Nanoleaf {
    pub fn new(host: &str, token: &str, effects: &[&str]) -> Arc<Nanoleaf> {
        Arc::new(Nanoleaf {
            client: crate::device::http_client(),
            base: format!("http://{}:16021/api/v1/{}", host, token),
            effects: effects.iter().map(|e| e.to_string()).collect(),
        })
    }

    async fn get(&self, path: &str) -> Value {
        let response = match self.client.get(format!("{}{}", self.base, path)).send().await {
            Ok(response) => response.json().await,
            Err(e) => Err(e),
        };

        response.unwrap_or_else(|e| {
            eprintln!("nanoleaf: {}", e);
            Value::Null
        })
    }

    // writes are answered with an empty 204 (an empty 400 when refused)
    async fn put(&self, path: &str, payload: Value) {
        if let Err(e) = self.client.put(format!("{}{}", self.base, path)).json(&payload).send().await {
            eprintln!("nanoleaf: {}", e);
        }
    }
}

// {"value": 120, "max": 360, "min": 0} -> 0.333
fn ratio(attribute: &Value) -> Number {
    let value = attribute["value"].as_f64().unwrap_or(0.0) / attribute["max"].as_f64().unwrap_or(1.0);
    Number::from_f64(value).unwrap_or_else(|| Number::from(0))
}

#[async_trait]
impl Device for Nanoleaf {
    fn capabilities(&self) -> Capabilities {
        Capabilities::COLOR
    }

    fn effects(&self) -> &[String] {
        &self.effects
    }

    async fn read(&self) -> Report {
        let state = self.get("/state").await;
        let effect = self.get("/effects/select").await;

        Report {
            on: Some(state["on"]["value"].as_bool().unwrap_or(false)),
            allow_on: Some(true),
            color: Some(ColorReport::Stored {
                hue: ratio(&state["hue"]),
                brightness: ratio(&state["brightness"]),
                saturation: ratio(&state["sat"]),
            }),
            // *Solid*, *Dynamic* and effects picked in the nanoleaf app are not ours
            effect: Some(effect.as_str().filter(|e| self.effects.iter().any(|known| known == e)).map(String::from)),
            ..Report::full()
        }
    }

    async fn apply(&self, cmd: &Command) -> Report {
        let color = cmd.color.map(ColorReport::Applied);

        match &cmd.effect {
            // the scenes send the same effect to every panel, and every panel
            // has a different set: not having it is not an error
            Some(Effect::Named(name)) if !self.effects.is_empty() && !self.effects.contains(name) => Report::full(),

            Some(Effect::Named(name)) => {
                self.put("/effects", json!({"select": name})).await;

                Report { on: Some(true), color, effect: Some(Some(name.clone())), ..Report::full() }
            }

            Some(Effect::Highlight(around)) => {
                let palette: Vec<Value> = (1..=3)
                    .map(|i| {
                        let [hue, saturation, brightness] = Hsb {
                            hue: around.hue,
                            saturation: around.saturation.map(|s| s / i as f64),
                            brightness: Some(1.0),
                        }
                        .scaled(HOMEKIT);

                        let mut entry = Map::new();
                        for (key, value) in [("hue", hue), ("saturation", saturation), ("brightness", brightness)] {
                            if let Some(value) = value {
                                entry.insert(key.into(), json!(value));
                            }
                        }
                        Value::Object(entry)
                    })
                    .collect();

                self.put(
                    "/effects",
                    json!({"write": {
                        "command": "display",
                        "animName": "New animation",
                        "animType": "highlight",
                        "colorType": "HSB",
                        "animData": null,
                        "palette": palette,
                        "brightnessRange": {"minValue": 50, "maxValue": 100},
                        "transTime": {"minValue": 5, "maxValue": 10},
                        "delayTime": {"minValue": 5, "maxValue": 10},
                        "loop": true,
                    }}),
                )
                .await;

                Report { on: Some(true), color: Some(ColorReport::Applied(*around)), effect: Some(None), ..Report::full() }
            }

            None => {
                let mut request = Map::new();

                if let Some(on) = cmd.on {
                    request.insert("on".into(), json!({"value": on}));
                }

                // a colour would switch the panel on again
                if let (Some(color), true) = (&cmd.color, cmd.on != Some(false)) {
                    for (key, value) in ["hue", "sat", "brightness"].into_iter().zip(color.scaled(HOMEKIT)) {
                        if let Some(value) = value {
                            request.insert(key.into(), json!({"value": value}));
                        }
                    }
                }

                // nothing to say (a "stop" sent to everything): an empty put makes the panel answer garbage
                if !request.is_empty() {
                    self.put("/state", Value::Object(request)).await;
                }

                // a colour leaves the panel in solid mode: no effect running
                Report { on: cmd.on, color, effect: Some(None), ..Report::full() }
            }
        }
    }

    async fn is_on(&self) -> bool {
        self.get("/state").await["on"]["value"].as_bool().unwrap_or(false)
    }
}
